File: scripts/dida365_lazy.py
#!/usr/bin/env python3
"""Lazy Dida365/TickTick MCP wrapper for Hermes Agent.

This script calls the official Dida365/TickTick MCP server from a short-lived
subprocess so the main Hermes Agent session can keep the native MCP server
configured but disabled. It is intended to be used from the dida365-lazy-mcp
skill.

Examples:
  python dida365_lazy.py list_projects
  python dida365_lazy.py create_task --args-json '{"task":{"title":"Test","projectId":"inbox"}}'
  python dida365_lazy.py complete_task --args-json '{"project_id":"inbox","task_id":"..."}'
"""
from __future__ import annotations

import argparse
import copy
import json
import os
import re
import sys
from pathlib import Path
from typing import Any

HERMES_HOME = Path(os.environ.get("HERMES_HOME", str(Path.home() / ".hermes"))).expanduser()
HERMES_REPO = Path(os.environ.get("HERMES_AGENT_REPO", str(HERMES_HOME / "hermes-agent"))).expanduser()
if HERMES_REPO.exists() and str(HERMES_REPO) not in sys.path:
    sys.path.insert(0, str(HERMES_REPO))

DEFAULT_SERVER_CANDIDATES = ("dida365", "ticktick")
DEFAULT_ALLOWED_TOOLS = {
    "list_projects",
    "create_task",
    "batch_add_tasks",
    "search_task",
    "get_task_by_id",
    "update_task",
    "complete_task",
    "complete_tasks_in_project",
    "list_undone_tasks_by_time_query",
    "list_undone_tasks_by_date",
    "get_project_with_undone_tasks",
}

SENSITIVE_RE = re.compile(
    r"(bearer\s+)[a-z0-9._~+\-/]+=*|"
    r"(token|api[_-]?key|secret|password|authorization)(\s*[=:]\s*)[^\s,}\]]+",
    re.IGNORECASE,
)


def redact(text: str) -> str:
    """Redact common credential-like substrings in error text."""

    def repl(match: re.Match[str]) -> str:
        if match.group(1):
            return match.group(1) + "[REDACTED]"
        return f"{match.group(2)}{match.group(3)}[REDACTED]"

    return SENSITIVE_RE.sub(repl, text)


def json_print(obj: Any) -> None:
    print(json.dumps(obj, ensure_ascii=False, separators=(",", ":")))


def parse_jsonish(text: Any) -> Any:
    """Parse JSON, including concatenated JSON objects returned as text."""
    if not isinstance(text, str):
        return text
    s = text.strip()
    if not s:
        return s
    try:
        return json.loads(s)
    except Exception:
        pass

    decoder = json.JSONDecoder()
    items = []
    idx = 0
    n = len(s)
    while idx < n:
        while idx < n and s[idx].isspace():
            idx += 1
        if idx >= n:
            break
        try:
            obj, end = decoder.raw_decode(s, idx)
        except Exception:
            return text
        items.append(obj)
        idx = end
    return items if items else text


def load_args(args_json: str | None, args_file: str | None) -> dict:
    if args_json and args_file:
        raise SystemExit("Use only one of --args-json or --args-file")
    if args_file:
        text = Path(args_file).expanduser().read_text(encoding="utf-8")
    elif args_json:
        text = args_json
    else:
        return {}
    try:
        value = json.loads(text)
    except Exception as exc:
        raise SystemExit(f"Invalid JSON args: {exc}") from exc
    if not isinstance(value, dict):
        raise SystemExit("Tool args must be a JSON object")
    return value


def strip_task(task: Any) -> Any:
    if not isinstance(task, dict):
        return task
    keys = [
        "id",
        "taskId",
        "projectId",
        "title",
        "content",
        "desc",
        "status",
        "priority",
        "startDate",
        "dueDate",
        "completedTime",
        "timeZone",
        "isAllDay",
        "kind",
        "url",
        "tags",
        "repeatFlag",
        "items",
    ]
    out = {k: task.get(k) for k in keys if k in task and task.get(k) is not None}
    if isinstance(out.get("items"), list):
        out["items"] = [
            {
                kk: item.get(kk)
                for kk in ["id", "title", "status", "startDate", "dueDate", "isAllDay"]
                if isinstance(item, dict) and kk in item
            }
            for item in out["items"][:20]
        ]
        if len(task.get("items") or []) > 20:
            out["items_truncated"] = len(task.get("items") or []) - 20
    return out


def strip_project(project: Any) -> Any:
    if not isinstance(project, dict):
        return project
    return {
        k: project.get(k)
        for k in ["id", "name", "kind", "viewMode", "closed", "groupId"]
        if k in project
    }


def generic_compact(value: Any, *, limit: int = 80, depth: int = 0) -> Any:
    if depth > 6:
        return "[TRUNCATED_DEPTH]"
    if isinstance(value, dict):
        return {
            str(k): generic_compact(v, limit=limit, depth=depth + 1)
            for k, v in list(value.items())[:120]
        }
    if isinstance(value, list):
        items = [generic_compact(v, limit=limit, depth=depth + 1) for v in value[:limit]]
        if len(value) > limit:
            items.append({"truncated": len(value) - limit})
        return items
    if isinstance(value, str) and len(value) > 2000:
        return value[:2000] + f"...[TRUNCATED {len(value) - 2000} chars]"
    return value


def _unwrap_payload(payload: Any) -> Any:
    if isinstance(payload, dict):
        return payload.get("task") or payload.get("data") or payload.get("result") or payload
    return payload


def _tasks_from_payload(payload: Any) -> list | None:
    if isinstance(payload, list):
        return payload
    if isinstance(payload, dict):
        for key in ("tasks", "items", "result", "data"):
            value = payload.get(key)
            if isinstance(value, list):
                return value
    return None


def compact_value(tool: str, payload: Any, limit: int = 80) -> Any:
    """Return compact, model-friendly JSON for common Dida365 payloads."""
    if tool == "list_projects":
        projects = payload if isinstance(payload, list) else payload.get("result", []) if isinstance(payload, dict) else []
        if isinstance(projects, list):
            return {"projects": [strip_project(p) for p in projects], "count": len(projects)}

    if tool in {"search_task", "list_undone_tasks_by_time_query", "list_undone_tasks_by_date"}:
        tasks = _tasks_from_payload(payload)
        if isinstance(tasks, list):
            return {
                "tasks": [strip_task(t) for t in tasks[:limit]],
                "count": len(tasks),
                "truncated": max(0, len(tasks) - limit),
            }

    if tool == "get_project_with_undone_tasks" and isinstance(payload, dict):
        project = payload.get("project") or payload.get("projectInfo") or payload
        tasks = payload.get("tasks") or payload.get("undoneTasks") or payload.get("result") or []
        return {
            "project": strip_project(project),
            "tasks": [strip_task(t) for t in tasks[:limit]] if isinstance(tasks, list) else tasks,
            "count": len(tasks) if isinstance(tasks, list) else None,
            "truncated": max(0, len(tasks) - limit) if isinstance(tasks, list) else 0,
        }

    if tool in {"get_task_by_id", "create_task", "update_task", "complete_task"}:
        inner = _unwrap_payload(payload)
        if isinstance(inner, list):
            return {"items": [strip_task(x) for x in inner[:limit]], "count": len(inner), "truncated": max(0, len(inner) - limit)}
        return strip_task(inner)

    if tool in {"batch_add_tasks", "complete_tasks_in_project"}:
        tasks = _tasks_from_payload(payload)
        if isinstance(tasks, list):
            return {"items": [strip_task(x) for x in tasks[:limit]], "count": len(tasks), "truncated": max(0, len(tasks) - limit)}
        return generic_compact(payload, limit=limit)

    return generic_compact(payload, limit=limit)


def load_hermes_config() -> dict:
    try:
        from hermes_cli.config import load_config
    except Exception as exc:
        raise SystemExit(f"Failed to import Hermes config loader from {HERMES_REPO}: {exc}") from exc
    return load_config() or {}


def select_server_config(cfg: dict, requested_server: str | None) -> tuple[str, dict]:
    servers = cfg.get("mcp_servers") or {}
    if not isinstance(servers, dict):
        raise SystemExit("No mcp_servers mapping found in Hermes config")

    if requested_server:
        server_cfg = servers.get(requested_server)
        if not isinstance(server_cfg, dict):
            raise SystemExit(f"No mcp_servers.{requested_server} config found in Hermes config")
        return requested_server, copy.deepcopy(server_cfg)

    env_server = os.environ.get("DIDA365_MCP_SERVER") or os.environ.get("TICKTICK_MCP_SERVER")
    if env_server and isinstance(servers.get(env_server), dict):
        return env_server, copy.deepcopy(servers[env_server])

    for candidate in DEFAULT_SERVER_CANDIDATES:
        if isinstance(servers.get(candidate), dict):
            return candidate, copy.deepcopy(servers[candidate])

    for name, server_cfg in servers.items():
        if not isinstance(server_cfg, dict):
            continue
        url = str(server_cfg.get("url") or "")
        if "mcp.dida365.com" in url or "ticktick" in url.lower() or "dida365" in url.lower():
            return name, copy.deepcopy(server_cfg)

    raise SystemExit("No Dida365/TickTick MCP server config found. Expected mcp_servers.dida365 or a server URL containing mcp.dida365.com")


def get_server_config(server: str | None) -> tuple[str, dict]:
    cfg = load_hermes_config()
    server_name, server_cfg = select_server_config(cfg, server)

    # Lazy mode: the main Hermes process can keep this MCP server disabled.
    # This subprocess enables it only for one operation.
    server_cfg["enabled"] = True
    server_cfg.setdefault("auth", "oauth")

    # Keep resources/prompts out of this lazy path unless the user explicitly
    # configured them. They are not needed for task operations and can bloat
    # subprocess discovery output/logs.
    tools_cfg = server_cfg.setdefault("tools", {})
    if isinstance(tools_cfg, dict):
        tools_cfg.setdefault("include", sorted(DEFAULT_ALLOWED_TOOLS))
        tools_cfg.setdefault("resources", False)
        tools_cfg.setdefault("prompts", False)

    return server_name, server_cfg


def sanitize_component(value: str) -> str:
    try:
        from tools.mcp_tool import sanitize_mcp_name_component

        return sanitize_mcp_name_component(value)
    except Exception:
        return re.sub(r"[^A-Za-z0-9_]", "_", str(value or ""))


def validate_tool_allowed(tool: str, server_cfg: dict) -> dict | None:
    if tool not in DEFAULT_ALLOWED_TOOLS:
        return {"ok": False, "error": f"Tool not allowed by lazy wrapper: {tool}", "allowed": sorted(DEFAULT_ALLOWED_TOOLS)}

    tools_filter = server_cfg.get("tools") if isinstance(server_cfg.get("tools"), dict) else {}
    include = tools_filter.get("include") if isinstance(tools_filter, dict) else None
    if include and tool not in set(include):
        return {"ok": False, "error": f"Tool {tool!r} is not included in mcp_servers.<server>.tools.include", "include": include}
    return None


def call_mcp(server: str | None, tool: str, tool_args: dict) -> dict:
    try:
        server_name, server_cfg = get_server_config(server)
    except SystemExit as exc:
        return {"ok": False, "tool": tool, "error": str(exc)}

    validation_error = validate_tool_allowed(tool, server_cfg)
    if validation_error:
        return validation_error

    from tools.mcp_tool import register_mcp_servers, shutdown_mcp_servers
    from tools.registry import registry

    try:
        names = register_mcp_servers({server_name: server_cfg})
        full_name = f"mcp_{sanitize_component(server_name)}_{sanitize_component(tool)}"
        if full_name not in set(names) and registry.get_entry(full_name) is None:
            available = [n for n in names if n.startswith(f"mcp_{sanitize_component(server_name)}_")]
            return {"ok": False, "tool": tool, "error": f"MCP tool not registered: {full_name}", "registered": available}

        raw = registry.dispatch(full_name, tool_args)
        try:
            outer = json.loads(raw) if isinstance(raw, str) else raw
        except Exception:
            outer = {"result": raw}

        if isinstance(outer, dict) and outer.get("error"):
            return {"ok": False, "tool": tool, "server": server_name, "error": redact(str(outer.get("error")))}

        payload: Any
        if isinstance(outer, dict):
            structured = outer.get("structuredContent")
            if isinstance(structured, dict) and "result" in structured:
                payload = structured.get("result")
            elif "result" in outer:
                payload = parse_jsonish(outer.get("result"))
            else:
                payload = outer
        else:
            payload = outer
        return {"ok": True, "tool": tool, "server": server_name, "payload": payload}
    except Exception as exc:
        return {"ok": False, "tool": tool, "server": server_name, "error": redact(f"{type(exc).__name__}: {exc}")}
    finally:
        try:
            shutdown_mcp_servers()
        except Exception:
            pass


def main() -> int:
    parser = argparse.ArgumentParser(description="Lazy Dida365/TickTick MCP caller for Hermes Agent")
    parser.add_argument("tool", help="Dida365/TickTick MCP tool name without the mcp_<server>_ prefix")
    parser.add_argument("--server", help="Hermes mcp_servers key to use. Defaults to dida365, ticktick, or URL autodetect.")
    parser.add_argument("--args-json", help="JSON object passed to the MCP tool")
    parser.add_argument("--args-file", help="Path to JSON file passed to the MCP tool")
    parser.add_argument("--raw", action="store_true", help="Return un-compacted payload")
    parser.add_argument("--dry-run", action="store_true", help="Validate args and config only; do not call MCP")
    parser.add_argument("--limit", type=int, default=80, help="Max list items in compact output")
    ns = parser.parse_args()

    tool_args = load_args(ns.args_json, ns.args_file)
    if ns.dry_run:
        try:
            server_name, cfg = get_server_config(ns.server)
            validation_error = validate_tool_allowed(ns.tool, cfg)
            transport = "http" if cfg.get("url") else "stdio" if cfg.get("command") else "unknown"
        except SystemExit as exc:
            json_print({"ok": False, "tool": ns.tool, "dry_run": True, "error": str(exc)})
            return 1
        if validation_error:
            json_print({**validation_error, "tool": ns.tool, "dry_run": True, "server": server_name})
            return 2
        json_print({"ok": True, "dry_run": True, "tool": ns.tool, "server": server_name, "args": tool_args, "configured": True, "transport": transport})
        return 0

    result = call_mcp(ns.server, ns.tool, tool_args)
    if result.get("ok") and not ns.raw:
        result = {"ok": True, "tool": ns.tool, "server": result.get("server"), "data": compact_value(ns.tool, result.get("payload"), limit=ns.limit)}
    elif result.get("ok") and ns.raw:
        result = {"ok": True, "tool": ns.tool, "server": result.get("server"), "data": result.get("payload")}
    json_print(result)
    return 0 if result.get("ok") else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/js/dida365_lazy.mjs
#!/usr/bin/env node
/**
 * Node.js lazy Dida365/TickTick MCP wrapper for Hermes Agent.
 *
 * This implementation is useful for users who prefer a Node runtime or want a
 * wrapper that does not import Hermes' Python modules. It still expects either
 * a Hermes-style MCP config/token store or an explicit access token via env.
 */
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import YAML from 'yaml';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';

const DEFAULT_SERVER_CANDIDATES = ['dida365', 'ticktick'];
const DEFAULT_MCP_URL = 'https://mcp.dida365.com';
const DEFAULT_ALLOWED_TOOLS = new Set([
  'list_projects',
  'create_task',
  'batch_add_tasks',
  'search_task',
  'get_task_by_id',
  'update_task',
  'complete_task',
  'complete_tasks_in_project',
  'list_undone_tasks_by_time_query',
  'list_undone_tasks_by_date',
  'get_project_with_undone_tasks',
]);

const __filename = fileURLToPath(import.meta.url);
const JS_DIR = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(JS_DIR, '..', '..');
const HERMES_HOME = path.resolve(process.env.HERMES_HOME || path.join(os.homedir(), '.hermes'));
const XDG_CONFIG_HOME = process.env.XDG_CONFIG_HOME || path.join(os.homedir(), '.config');

function loadDotEnvFile(filePath) {
  if (!fs.existsSync(filePath)) return;
  const lines = fs.readFileSync(filePath, 'utf8').split(/\r?\n/);
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const idx = trimmed.indexOf('=');
    if (idx <= 0) continue;
    const key = trimmed.slice(0, idx).trim();
    if (process.env[key] !== undefined) continue;
    let value = trimmed.slice(idx + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    process.env[key] = value;
  }
}

loadDotEnvFile(path.join(PROJECT_ROOT, '.env'));

function jsonPrint(obj) {
  process.stdout.write(`${JSON.stringify(obj)}\n`);
}

function redact(text) {
  return String(text)
    .replace(/(bearer\s+)[a-z0-9._~+\-/]+=*/gi, '$1[REDACTED]')
    .replace(/(token|api[_-]?key|secret|password|authorization)(\s*[=:]\s*)[^\s,}\]]+/gi, '$1$2[REDACTED]');
}

function parseArgs(argv) {
  const out = {
    tool: null,
    server: null,
    argsJson: null,
    argsFile: null,
    raw: false,
    dryRun: false,
    limit: 80,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === '--server') out.server = argv[++i];
    else if (arg === '--args-json') out.argsJson = argv[++i];
    else if (arg === '--args-file') out.argsFile = argv[++i];
    else if (arg === '--raw') out.raw = true;
    else if (arg === '--dry-run') out.dryRun = true;
    else if (arg === '--limit') out.limit = Number(argv[++i] || '80');
    else if (arg === '-h' || arg === '--help') {
      out.help = true;
    } else if (!out.tool) {
      out.tool = arg;
    } else {
      throw new Error(`Unexpected argument: ${arg}`);
    }
  }
  return out;
}

function usage() {
  return `Usage: dida365_lazy.mjs <tool> [--server dida365] [--args-json JSON | --args-file FILE] [--dry-run] [--raw] [--limit N]\n`;
}

function loadToolArgs(argsJson, argsFile) {
  if (argsJson && argsFile) throw new Error('Use only one of --args-json or --args-file');
  if (!argsJson && !argsFile) return {};
  const text = argsFile ? fs.readFileSync(path.resolve(argsFile), 'utf8') : argsJson;
  const value = JSON.parse(text);
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('Tool args must be a JSON object');
  }
  return value;
}

function loadHermesConfig() {
  const configPath = process.env.HERMES_CONFIG || path.join(HERMES_HOME, 'config.yaml');
  if (!fs.existsSync(configPath)) {
    return null;
  }
  const text = fs.readFileSync(configPath, 'utf8');
  return YAML.parse(text) || {};
}

function standaloneServerConfig(requestedServer) {
  const name = requestedServer || process.env.DIDA365_MCP_SERVER || process.env.TICKTICK_MCP_SERVER || 'dida365';
  const url = process.env.DIDA365_MCP_URL || process.env.TICKTICK_MCP_URL || process.env.MCP_SERVER_URL || DEFAULT_MCP_URL;
  return [name, {
    url,
    auth: 'oauth',
    enabled: true,
    tools: {
      include: Array.from(DEFAULT_ALLOWED_TOOLS).sort(),
      resources: false,
      prompts: false,
    },
  }];
}

function selectServerConfig(config, requestedServer) {
  const servers = config.mcp_servers || {};
  if (!servers || typeof servers !== 'object') throw new Error('No mcp_servers mapping found in Hermes config');

  if (requestedServer) {
    if (!servers[requestedServer] || typeof servers[requestedServer] !== 'object') {
      throw new Error(`No mcp_servers.${requestedServer} config found in Hermes config`);
    }
    return [requestedServer, structuredClone(servers[requestedServer])];
  }

  const envServer = process.env.DIDA365_MCP_SERVER || process.env.TICKTICK_MCP_SERVER;
  if (envServer && servers[envServer] && typeof servers[envServer] === 'object') {
    return [envServer, structuredClone(servers[envServer])];
  }

  for (const candidate of DEFAULT_SERVER_CANDIDATES) {
    if (servers[candidate] && typeof servers[candidate] === 'object') {
      return [candidate, structuredClone(servers[candidate])];
    }
  }

  for (const [name, serverConfig] of Object.entries(servers)) {
    if (!serverConfig || typeof serverConfig !== 'object') continue;
    const url = String(serverConfig.url || '').toLowerCase();
    if (url.includes('mcp.dida365.com') || url.includes('dida365') || url.includes('ticktick')) {
      return [name, structuredClone(serverConfig)];
    }
  }
  throw new Error('No Dida365/TickTick MCP server config found. Expected mcp_servers.dida365 or a server URL containing mcp.dida365.com');
}

function getServerConfig(serverName) {
  const config = loadHermesConfig();
  if (!config || !config.mcp_servers) {
    return standaloneServerConfig(serverName);
  }
  const [name, serverConfig] = selectServerConfig(config, serverName);
  serverConfig.enabled = true;
  serverConfig.auth ||= 'oauth';
  serverConfig.tools ||= {};
  if (!serverConfig.tools.include) serverConfig.tools.include = Array.from(DEFAULT_ALLOWED_TOOLS).sort();
  if (serverConfig.tools.resources === undefined) serverConfig.tools.resources = false;
  if (serverConfig.tools.prompts === undefined) serverConfig.tools.prompts = false;
  return [name, serverConfig];
}

function validateToolAllowed(tool, serverConfig) {
  if (!DEFAULT_ALLOWED_TOOLS.has(tool)) {
    return { ok: false, error: `Tool not allowed by lazy wrapper: ${tool}`, allowed: Array.from(DEFAULT_ALLOWED_TOOLS).sort() };
  }
  const include = serverConfig.tools && Array.isArray(serverConfig.tools.include) ? serverConfig.tools.include : null;
  if (include && include.length && !new Set(include).has(tool)) {
    return { ok: false, error: `Tool '${tool}' is not included in mcp_servers.<server>.tools.include`, include };
  }
  return null;
}

function loadAccessToken(serverName) {
  const explicit = process.env.DIDA365_MCP_ACCESS_TOKEN || process.env.TICKTICK_MCP_ACCESS_TOKEN || process.env.MCP_ACCESS_TOKEN;
  if (explicit) return { accessToken: explicit, source: 'env' };

  const explicitPath = process.env.DIDA365_MCP_TOKEN_FILE;
  const tokenPaths = [
    explicitPath,
    path.join(XDG_CONFIG_HOME, 'dida365-lazy-mcp', 'tokens', `${serverName}.json`),
    path.join(HERMES_HOME, 'mcp-tokens', `${serverName}.json`),
  ].filter(Boolean);

  for (const tokenPath of tokenPaths) {
    if (!fs.existsSync(tokenPath)) continue;
    const token = JSON.parse(fs.readFileSync(tokenPath, 'utf8'));
    const accessToken = token.access_token;
    if (!accessToken) return { accessToken: null, source: tokenPath, warning: `No access_token in token file: ${tokenPath}` };
    if (token.expires_at && Number(token.expires_at) < Date.now() / 1000) {
      return {
        accessToken,
        source: tokenPath,
        warning: 'Access token appears expired. Run a compatible OAuth refresh/login flow or provide DIDA365_MCP_ACCESS_TOKEN after refreshing authorization.',
      };
    }
    return { accessToken, source: tokenPath };
  }

  return { accessToken: null, source: tokenPaths[tokenPaths.length - 1], warning: `No access token available. Set DIDA365_MCP_ACCESS_TOKEN or run a compatible OAuth login flow.` };
}

function parseJsonish(value) {
  if (typeof value !== 'string') return value;
  const trimmed = value.trim();
  if (!trimmed) return trimmed;
  try {
    return JSON.parse(trimmed);
  } catch (_) {
    return value;
  }
}

function stripProject(project) {
  if (!project || typeof project !== 'object' || Array.isArray(project)) return project;
  const keys = ['id', 'name', 'kind', 'viewMode', 'closed', 'groupId'];
  return Object.fromEntries(keys.filter((k) => k in project).map((k) => [k, project[k]]));
}

function stripTask(task) {
  if (!task || typeof task !== 'object' || Array.isArray(task)) return task;
  const keys = [
    'id', 'taskId', 'projectId', 'title', 'content', 'desc', 'status', 'priority',
    'startDate', 'dueDate', 'completedTime', 'timeZone', 'isAllDay', 'kind', 'url',
    'tags', 'repeatFlag', 'items',
  ];
  const out = {};
  for (const key of keys) {
    if (task[key] !== undefined && task[key] !== null) out[key] = task[key];
  }
  if (Array.isArray(out.items)) {
    out.items = out.items.slice(0, 20).map((item) => {
      if (!item || typeof item !== 'object') return item;
      const itemKeys = ['id', 'title', 'status', 'startDate', 'dueDate', 'isAllDay'];
      return Object.fromEntries(itemKeys.filter((k) => k in item).map((k) => [k, item[k]]));
    });
    if (task.items.length > 20) out.items_truncated = task.items.length - 20;
  }
  return out;
}

function genericCompact(value, limit = 80, depth = 0) {
  if (depth > 6) return '[TRUNCATED_DEPTH]';
  if (Array.isArray(value)) {
    const items = value.slice(0, limit).map((v) => genericCompact(v, limit, depth + 1));
    if (value.length > limit) items.push({ truncated: value.length - limit });
    return items;
  }
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value).slice(0, 120).map(([k, v]) => [k, genericCompact(v, limit, depth + 1)]));
  }
  if (typeof value === 'string' && value.length > 2000) return `${value.slice(0, 2000)}...[TRUNCATED ${value.length - 2000} chars]`;
  return value;
}

function unwrapPayload(payload) {
  if (payload && typeof payload === 'object' && !Array.isArray(payload)) {
    return payload.task || payload.data || payload.result || payload;
  }
  return payload;
}

function tasksFromPayload(payload) {
  if (Array.isArray(payload)) return payload;
  if (payload && typeof payload === 'object') {
    for (const key of ['tasks', 'items', 'result', 'data']) {
      if (Array.isArray(payload[key])) return payload[key];
    }
  }
  return null;
}

function compactValue(tool, payload, limit = 80) {
  if (tool === 'list_projects') {
    const projects = Array.isArray(payload) ? payload : (payload && Array.isArray(payload.result) ? payload.result : []);
    return { projects: projects.map(stripProject), count: projects.length };
  }
  if (['search_task', 'list_undone_tasks_by_time_query', 'list_undone_tasks_by_date'].includes(tool)) {
    const tasks = tasksFromPayload(payload);
    if (Array.isArray(tasks)) {
      return { tasks: tasks.slice(0, limit).map(stripTask), count: tasks.length, truncated: Math.max(0, tasks.length - limit) };
    }
  }
  if (tool === 'get_project_with_undone_tasks' && payload && typeof payload === 'object' && !Array.isArray(payload)) {
    const project = payload.project || payload.projectInfo || payload;
    const tasks = payload.tasks || payload.undoneTasks || payload.result || [];
    return {
      project: stripProject(project),
      tasks: Array.isArray(tasks) ? tasks.slice(0, limit).map(stripTask) : tasks,
      count: Array.isArray(tasks) ? tasks.length : null,
      truncated: Array.isArray(tasks) ? Math.max(0, tasks.length - limit) : 0,
    };
  }
  if (['get_task_by_id', 'create_task', 'update_task', 'complete_task'].includes(tool)) {
    const inner = unwrapPayload(payload);
    if (Array.isArray(inner)) return { items: inner.slice(0, limit).map(stripTask), count: inner.length, truncated: Math.max(0, inner.length - limit) };
    return stripTask(inner);
  }
  if (['batch_add_tasks', 'complete_tasks_in_project'].includes(tool)) {
    const tasks = tasksFromPayload(payload);
    if (Array.isArray(tasks)) return { items: tasks.slice(0, limit).map(stripTask), count: tasks.length, truncated: Math.max(0, tasks.length - limit) };
  }
  return genericCompact(payload, limit);
}

function extractPayload(result) {
  if (result && typeof result === 'object') {
    if (result.structuredContent && typeof result.structuredContent === 'object') {
      if ('result' in result.structuredContent) return result.structuredContent.result;
      return result.structuredContent;
    }
    if (Array.isArray(result.content)) {
      const textItems = result.content.filter((item) => item && item.type === 'text' && typeof item.text === 'string');
      if (textItems.length === 1) return parseJsonish(textItems[0].text);
      if (textItems.length > 1) return textItems.map((item) => parseJsonish(item.text));
    }
    if ('result' in result) return parseJsonish(result.result);
  }
  return result;
}

async function callMcp(serverName, serverConfig, tool, toolArgs) {
  if (!serverConfig.url) throw new Error('The JS wrapper currently supports HTTP/StreamableHTTP MCP servers only; mcp_servers.<server>.url is required.');
  const tokenInfo = loadAccessToken(serverName);
  if (!tokenInfo.accessToken) throw new Error(tokenInfo.warning || 'No access token available');

  const headers = {
    ...(serverConfig.headers || {}),
    Authorization: `Bearer ${tokenInfo.accessToken}`,
  };

  const transport = new StreamableHTTPClientTransport(new URL(serverConfig.url), {
    requestInit: { headers },
  });
  const client = new Client({ name: 'dida365-lazy-mcp-js', version: '1.1.0' });
  try {
    await client.connect(transport);
    const listed = await client.listTools();
    const toolNames = new Set((listed.tools || []).map((t) => t.name));
    if (!toolNames.has(tool)) {
      return { ok: false, tool, server: serverName, error: `MCP tool not available: ${tool}`, available: Array.from(toolNames).filter((name) => DEFAULT_ALLOWED_TOOLS.has(name)).sort() };
    }
    const result = await client.callTool({ name: tool, arguments: toolArgs });
    if (result && result.isError) {
      return { ok: false, tool, server: serverName, error: redact(JSON.stringify(result)) };
    }
    return { ok: true, tool, server: serverName, payload: extractPayload(result), warning: tokenInfo.warning };
  } finally {
    try { await client.close(); } catch (_) {}
    try { await transport.close(); } catch (_) {}
  }
}

async function main() {
  const ns = parseArgs(process.argv.slice(2));
  if (ns.help || !ns.tool) {
    process.stdout.write(usage());
    return ns.help ? 0 : 2;
  }
  const toolArgs = loadToolArgs(ns.argsJson, ns.argsFile);
  const [serverName, serverConfig] = getServerConfig(ns.server);
  const validation = validateToolAllowed(ns.tool, serverConfig);
  if (validation) {
    jsonPrint({ ...validation, tool: ns.tool, server: serverName });
    return 2;
  }
  if (ns.dryRun) {
    jsonPrint({ ok: true, dry_run: true, tool: ns.tool, server: serverName, args: toolArgs, configured: true, transport: serverConfig.url ? 'http' : 'stdio' });
    return 0;
  }
  const result = await callMcp(serverName, serverConfig, ns.tool, toolArgs);
  if (result.ok && !ns.raw) {
    const out = { ok: true, tool: ns.tool, server: result.server, data: compactValue(ns.tool, result.payload, ns.limit) };
    if (result.warning) out.warning = result.warning;
    jsonPrint(out);
  } else if (result.ok && ns.raw) {
    const out = { ok: true, tool: ns.tool, server: result.server, data: result.payload };
    if (result.warning) out.warning = result.warning;
    jsonPrint(out);
  } else {
    jsonPrint(result);
  }
  return result.ok ? 0 : 1;
}

main().then((code) => process.exit(code)).catch((err) => {
  jsonPrint({ ok: false, error: redact(`${err.name || 'Error'}: ${err.message || err}`) });
  process.exit(1);
});
